File: matlab/src/api/pfqn/pfqn_nre.m
%{
%{
 % @file pfqn_nre.m
 % @brief Normalizing constant via saddle-tilted Edgeworth (NRE) approximation.
%}
%}

%{
%{
 % @brief Normalizing constant via saddle-tilted Edgeworth (NRE) approximation.
 %
 % Evaluates the Norlund-Rice integral form of the limited-load-dependent
 % normalizing constant (Casale-Harrison-Ong, Perform. Eval. 152, 2021,
 % Thm. 6) by steepest descent instead of a Laplace approximation on the
 % untilted contour, as done by pfqn_nrl and pfqn_nrp. Two corrections are
 % applied over those methods:
 %
 %   1. The integrand is invariant under t -> t + c*1, since h is homogeneous
 %      of degree sum(N) in the class variables and that degree cancels
 %      against exp(-1i*N*t). The redundant direction is quotiented out, so
 %      the integral is (R-1)-dimensional, not R-dimensional. pfqn_nrl and
 %      pfqn_nrp integrate over R dimensions and let the logit/probit
 %      Jacobian supply curvature along the null direction, which is an
 %      artifact of the change of variables rather than of the integrand.
 %   2. The contour radii are tilted per class to the saddle point, i.e. to
 %      the X solving X_r*dlog(h)/dX_r = N_r, so that the origin is a
 %      stationary point of the phase. On the untilted contour X = 1 it is
 %      not, which is the leading source of bias in pfqn_nrl and pfqn_nrp.
 %
 % A second-order Edgeworth term built from the third and fourth cumulants
 % of the tilted distribution is then added, giving a relative error of
 % O(1/sum(N)^2) instead of O(1) for a heuristic Gaussian fit.
 %
 % All integrand evaluations are at real positive demands, so unlike
 % pfqn_nrl and pfqn_nrp this method needs no complex arithmetic and runs
 % entirely in the log domain through pfqn_lldsingle, whose cost is linear
 % rather than quadratic in the population wherever the rates settle.
 %
 % Cost is O(I*R^2 + R^4) evaluations of a single-class LLD normalizing
 % constant, hence polynomial in the number of classes.
 %
 % @fn pfqn_nre(L, N, Z, alpha, options)
 % @param L Service demand matrix (MxR).
 % @param N Population vector (1xR).
 % @param Z Think time vector (1xR or DxR).
 % @param alpha Load-dependent rate matrix (Mx sum(N)).
 % @param options Solver options.
 % @param vfix Optional tilt to use instead of solving the saddle-point
 %        equation. Supplying the tilt obtained at a nearby population makes
 %        numerator and denominator of a ratio share one expansion point, which
 %        is the Tierney-Kadane arrangement; omit it for the standard estimator.
 % @return lG Logarithm of normalizing constant.
 % @return lGs Logarithm of the saddlepoint term alone, i.e. of lG with the
 %         Edgeworth factor omitted, so that lG-lGs is the correction.
 % @return vsad The tilt actually used, for reuse at a nearby population.
 % @return G Normalizing constant.
%}
%}
function [lG,G,lGs,vsad] = pfqn_nre(L,N,Z,alpha,options,vfix)
N = N(:)';
if sum(N)<0
    lG = -Inf; G = 0; lGs = lG; vsad = [];
    return
end
if sum(N)==0
    lG = 0; G = 1; lGs = lG; vsad = [];
    return
end
if nargin<3
    Z = zeros(1,length(N));
end
if nargin<5 || isempty(options)
    options = SolverNC.defaultOptions;
end
Nt = sum(N);
if size(alpha,2) < Nt
    line_error(mfilename,'the load-dependent rate matrix must have at least sum(N) columns.');
end
alpha = alpha(:,1:Nt); % trim so that every rate used downstream is positive
if ~isempty(Z) && sum(Z(:))>0
    L = [L; sum(Z,1)];
    alpha(end+1,1:Nt) = 1:Nt; % the delay is an infinite server station
end
[M,R] = size(L);
if M==1
    [~,lG] = pfqn_gld(L,N,alpha,options);
    G = exp(lG); lGs = lG; vsad = [];
    return
end

% scale demands in [0,1] per class, the residual factor is exact by homogeneity
Lmax = max(L,[],1);
Lmax(Lmax<=0) = 1;
L = L ./ repmat(Lmax,M,1);
lGscale = N*log(Lmax)';

if R==1
    % coefficient extraction is the identity in a single class
    lG = pfqn_lldsingle(L,Nt,alpha,options) + lGscale;
    G = exp(lG); lGs = lG; vsad = [];
    return
end

d = R - 1; % dimension of the quotient torus

hstep = 2e-2;    % finite-difference step, results are flat over [5e-3,5e-2]
% The stencil is memoised on the offset, which stays within [-4,4] along each
% dimension and leaves the origin in at most four dimensions at once, so the
% points it visits number O(d^4). Indexing them densely takes span^d entries,
% which is what used to stop the method at eight classes; a map costs space
% proportional to the work instead and leaves the arithmetic untouched.
mkcache = @() containers.Map('KeyType','char','ValueType','double');
cacheVal = mkcache();
vbase = zeros(1,d);
Nd = N(1:d);

% ---- saddle point: minimise the convex F(v) = K(v) - Nd*v ----
% A tilt supplied by the caller is used as given, so that a ratio of two
% constants can be expanded about one common point rather than two.
if nargin >= 6 && ~isempty(vfix)
    vbase = vfix(:)';
    converged = true;
else
converged = false;
for it = 1:100
    cacheVal = mkcache();
    grad = zeros(d,1);
    hess = zeros(d,d);
    for a = 1:d
        grad(a) = (cgf(unitoff(a,1)) - cgf(unitoff(a,-1)))/(2*hstep) - Nd(a);
    end
    for a = 1:d
        for b = 1:d
            hess(a,b) = (cgf(unitoff(a,1)+unitoff(b,1)) - cgf(unitoff(a,1)+unitoff(b,-1)) ...
                - cgf(unitoff(a,-1)+unitoff(b,1)) + cgf(unitoff(a,-1)+unitoff(b,-1)))/(4*hstep^2);
        end
    end
    step = -(hess\grad)';
    F0 = cgf(zeros(1,d)) - Nd*vbase';
    tau = 1;
    while tau > 1e-10
        vtry = vbase + tau*step;
        if cgfat(vtry) - Nd*vtry' <= F0
            break
        end
        tau = tau/2;
    end
    vbase = vbase + tau*step;
    % Newton converges to the root of the differenced gradient, whose own
    % O(hstep^2) bias puts any absolute gradient target out of reach
    if norm(tau*step) < 1e-10
        converged = true;
        break
    end
end
end
if ~converged
    line_warning(mfilename,'the saddle point search did not converge, the estimate may be inaccurate.');
end
vsad = vbase;

% ---- cumulants of the tilted distribution at the saddle ----
cacheVal = mkcache();
K0 = cgf(zeros(1,d));
Sigma = zeros(d,d);
for a = 1:d
    for b = 1:d
        Sigma(a,b) = (cgf(unitoff(a,1)+unitoff(b,1)) - cgf(unitoff(a,1)+unitoff(b,-1)) ...
            - cgf(unitoff(a,-1)+unitoff(b,1)) + cgf(unitoff(a,-1)+unitoff(b,-1)))/(4*hstep^2);
    end
end
Sigma = (Sigma + Sigma')/2;
if min(eig(Sigma)) <= 0
    line_error(mfilename,'the tilted covariance is singular, a class has no demand at any station.');
end

k3 = zeros(d,d,d);
for a = 1:d
    for b = 1:d
        for c = 1:d
            acc = 0;
            for s = 0:7
                sg = 1 - 2*[bitget(s,1),bitget(s,2),bitget(s,3)];
                acc = acc + prod(sg)*cgf(sg(1)*unitoff(a,1) + sg(2)*unitoff(b,1) + sg(3)*unitoff(c,1));
            end
            k3(a,b,c) = acc/(8*hstep^3);
        end
    end
end

k4 = zeros(d,d,d,d);
for a = 1:d
    for b = 1:d
        for c = 1:d
            for e = 1:d
                acc = 0;
                for s = 0:15
                    sg = 1 - 2*[bitget(s,1),bitget(s,2),bitget(s,3),bitget(s,4)];
                    acc = acc + prod(sg)*cgf(sg(1)*unitoff(a,1) + sg(2)*unitoff(b,1) ...
                        + sg(3)*unitoff(c,1) + sg(4)*unitoff(e,1));
                end
                k4(a,b,c,e) = acc/(16*hstep^4);
            end
        end
    end
end

% ---- second-order Edgeworth factor, see the derivation in the header ----
S = inv(Sigma);
rho4 = S(:)'*reshape(k4,[d*d,d*d])*S(:);
u = reshape(S(:)'*reshape(k3,[d*d,d]),1,d);
rhoA = u*S*u';
T = reshape(S*reshape(k3,d,d*d),[d d d]);
T = permute(reshape(S*reshape(permute(T,[2 1 3]),d,d*d),[d d d]),[2 1 3]);
T = permute(reshape(S*reshape(permute(T,[3 1 2]),d,d*d),[d d d]),[2 3 1]);
rhoB = sum(k3(:).*T(:));
corr = 1 + rho4/8 - (3*rhoA + 2*rhoB)/24;
if corr <= 0
    line_warning(mfilename,'the Edgeworth correction is non-positive, falling back on the saddlepoint term.');
    corr = 1;
end

lGs = K0 - Nd*vbase' - (d/2)*log(2*pi) - 0.5*log(det(Sigma)) + lGscale;
lG = lGs + log(corr);
G = exp(lG);

    function y = cgf(off)
        % cumulant generating function at vbase+off*hstep, memoised on the stencil
        key = char(off+5);   % offsets lie in [-4,4], so off+5 is a valid char code
        if isKey(cacheVal,key)
            y = cacheVal(key);
        else
            y = cgfat(vbase + off*hstep);
            cacheVal(key) = y;
        end
    end

    function y = cgfat(v)
        % log of the single-class LLD constant at the class tilt X=[exp(v),1]
        y = pfqn_lldsingle(L*[exp(v(:));1], Nt, alpha, options);
    end

    function off = unitoff(a,sgn)
        off = zeros(1,d);
        off(a) = sgn;
    end
end

File: jar/src/main/java/jline/api/pfqn/nc/Pfqn_nre.java
/**
 * @file Norlund-Rice Edgeworth (NRE) method for load-dependent normalizing constants
 *
 * Port of matlab/src/api/pfqn/pfqn_nre.m. Evaluates the Norlund-Rice integral
 * form of the limited load-dependent normalizing constant by steepest descent
 * on a saddle-tilted contour, with a second-order Edgeworth correction.
 *
 * @since LINE 3.0
 */
package jline.api.pfqn.nc;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

import org.apache.commons.math3.complex.Complex;
import org.apache.commons.math3.util.FastMath;

import jline.GlobalConstants;
import jline.api.pfqn.ld.Pfqn_gld;
import jline.api.pfqn.ld.Pfqn_lldsingle;
import jline.io.InputOutput;
import jline.solvers.SolverOptions;
import jline.util.matrix.Matrix;

/**
 * Normalizing constant via the saddle-tilted Edgeworth (NRE) approximation.
 *
 * Two corrections are applied over Pfqn_nrl and Pfqn_nrp: the integrand is
 * invariant under t -> t + c*1, so the redundant direction is quotiented out
 * and the integral is (R-1)-dimensional; and the contour radii are tilted per
 * class to the saddle point, making the origin a stationary point of the
 * phase. A second-order Edgeworth term built from the third and fourth
 * cumulants of the tilted distribution then gives a relative error of
 * O(1/sum(N)^2). Every integrand evaluation is at real positive demands, so
 * unlike Pfqn_nrl and Pfqn_nrp no complex arithmetic is involved.
 */
public final class Pfqn_nre {
    private Pfqn_nre() {}

    /** Finite-difference step; results are flat over [5e-3,5e-2]. */
    private static final double HSTEP = 2e-2;
    /** Stencil offsets stay within [-4,4] along each dimension. */
    private static final int SPAN = 9;
    /** Beyond 8 classes the fourth-cumulant tensor is no longer affordable. */
    private static final int MAX_DIM = 7;

    /**
     * MATLAB's {@code [lG,G,lGs,vsad]}: the constant, the saddlepoint term on
     * its own and the tilt the expansion was taken about.
     *
     * <p>{@code lG - lGs} is the Edgeworth correction, so a caller that wants
     * the plain saddlepoint estimate reads {@code lGs} rather than re-deriving
     * it. {@code vsad} is null on the shortcut arms that never solve a saddle
     * point, matching the reference's empty {@code []}.
     */
    public static final class Result {
        /** Logarithm of the normalizing constant, Edgeworth correction included. */
        public final double lG;
        /** The normalizing constant, {@code exp(lG)}. */
        public final double G;
        /** Logarithm of the saddlepoint term alone, i.e. lG with the Edgeworth factor omitted. */
        public final double lGs;
        /** The tilt actually used (1xd), null when no saddle point was solved. */
        public final Matrix vsad;

        Result(double lG, double lGs, Matrix vsad) {
            this.lG = lG;
            this.G = FastMath.exp(lG);
            this.lGs = lGs;
            this.vsad = vsad;
        }
    }

    /**
     * Logarithm of the normalizing constant of a limited load-dependent model.
     *
     * @param Lin     service demand matrix (MxR)
     * @param N       population vector (1xR)
     * @param Z       think time vector (1xR or DxR)
     * @param alphaIn load-dependent rate matrix (Mx sum(N))
     * @param options solver options
     * @return logarithm of the normalizing constant
     */
    public static double pfqn_nre(Matrix Lin, Matrix N, Matrix Z, Matrix alphaIn, SolverOptions options) {
        return pfqn_nre_full(Lin, N, Z, alphaIn, options, null).lG;
    }

    /**
     * The full form of the reference's four outputs, named alike in the native
     * python and C++ ports: the constant, the saddlepoint term alone and the tilt used.
     *
     * @param Lin     service demand matrix (MxR)
     * @param N       population vector (1xR)
     * @param Z       think time vector (1xR or DxR)
     * @param alphaIn load-dependent rate matrix (Mx sum(N))
     * @param options solver options
     * @param vfix    tilt to use instead of solving the saddle-point equation,
     *                null for the standard estimator. Supplying the tilt
     *                obtained at a nearby population makes numerator and
     *                denominator of a ratio share one expansion point, which is
     *                the Tierney-Kadane arrangement.
     * @return the constant, the saddlepoint term and the tilt
     */
    public static Result pfqn_nre_full(Matrix Lin, Matrix N, Matrix Z, Matrix alphaIn,
                                       SolverOptions options, Matrix vfix) {
        double Ntd = N.elementSum();
        if (Ntd < 0) {
            return new Result(GlobalConstants.NegInf, GlobalConstants.NegInf, null);
        }
        if (Ntd == 0.0) {
            return new Result(0.0, 0.0, null);
        }
        int Nt = (int) FastMath.round(Ntd);
        if (alphaIn.getNumCols() < Nt) {
            throw new RuntimeException("pfqn_nre: the load-dependent rate matrix must have at least sum(N) columns.");
        }
        // trim so that every rate used downstream is positive
        Matrix alpha = new Matrix(alphaIn.getNumRows(), Nt);
        for (int i = 0; i < alphaIn.getNumRows(); i++) {
            for (int j = 0; j < Nt; j++) {
                alpha.set(i, j, alphaIn.get(i, j));
            }
        }
        Matrix L = Lin.copy();
        if (Z != null && !Z.isEmpty() && Z.elementSum() > 0) {
            Matrix Zrow = Z.getNumRows() > 1 ? Z.sumCols() : Z;
            L = Matrix.concatRows(L, Zrow, null);
            Matrix delayRates = new Matrix(1, Nt);
            for (int j = 0; j < Nt; j++) {
                delayRates.set(0, j, j + 1.0); // the delay is an infinite server station
            }
            alpha = Matrix.concatRows(alpha, delayRates, null);
        }
        int M = L.getNumRows();
        int R = L.getNumCols();
        if (M == 1) {
            double lGone = Pfqn_gld.pfqn_gld(L, N, alpha, options).lG;
            return new Result(lGone, lGone, null);
        }

        // scale demands in [0,1] per class, the residual factor is exact by homogeneity
        double lGscale = 0.0;
        for (int r = 0; r < R; r++) {
            double colMax = 0.0;
            for (int i = 0; i < M; i++) {
                if (L.get(i, r) > colMax) {
                    colMax = L.get(i, r);
                }
            }
            if (colMax <= 0) {
                colMax = 1.0;
            }
            for (int i = 0; i < M; i++) {
                L.set(i, r, L.get(i, r) / colMax);
            }
            lGscale += N.get(r) * FastMath.log(colMax);
        }

        Matrix Ntmat = new Matrix(1, 1);
        Ntmat.set(0, 0, Nt);
        if (R == 1) {
            // coefficient extraction is the identity in a single class
            double lGone = Pfqn_lldsingle.pfqn_lldsingle(L, Ntmat, alpha, options).lG + lGscale;
            return new Result(lGone, lGone, null);
        }

        int d = R - 1; // dimension of the quotient torus
        if (d > MAX_DIM) {
            throw new RuntimeException("pfqn_nre: pfqn_nre is limited to 8 classes, use nrl or clw beyond that.");
        }
        double[] Nd = new double[d];
        for (int a = 0; a < d; a++) {
            Nd[a] = N.get(a);
        }
        Cgf cgf = new Cgf(L, Ntmat, alpha, options, d);

        // ---- saddle point: minimise the convex F(v) = K(v) - Nd*v ----
        // A tilt supplied by the caller is used as given, so that a ratio of two
        // constants can be expanded about one common point rather than two.
        double[] vbase = new double[d];
        boolean converged = false;
        boolean tiltGiven = vfix != null && !vfix.isEmpty();
        if (tiltGiven) {
            if (vfix.length() < d) {
                throw new RuntimeException("pfqn_nre: the supplied tilt must have one entry per "
                        + "quotient dimension (R-1).");
            }
            for (int a = 0; a < d; a++) {
                vbase[a] = vfix.get(a);
            }
            converged = true;
        }
        for (int it = 0; !tiltGiven && it < 100; it++) {
            cgf.reset(vbase);
            Matrix grad = new Matrix(d, 1);
            Matrix hess = new Matrix(d, d);
            for (int a = 0; a < d; a++) {
                grad.set(a, 0, (cgf.at(unitoff(d, a, 1)) - cgf.at(unitoff(d, a, -1))) / (2 * HSTEP) - Nd[a]);
            }
            for (int a = 0; a < d; a++) {
                for (int b = 0; b < d; b++) {
                    hess.set(a, b, secondDiff(cgf, d, a, b));
                }
            }
            Matrix step = hess.inv().mult(grad);
            double F0 = cgf.at(new int[d]) - dot(Nd, vbase);
            double tau = 1;
            double[] vtry = new double[d];
            while (tau > 1e-10) {
                for (int a = 0; a < d; a++) {
                    vtry[a] = vbase[a] - tau * step.get(a, 0);
                }
                if (cgf.atPoint(vtry) - dot(Nd, vtry) <= F0) {
                    break;
                }
                tau = tau / 2;
            }
            double stepNorm = 0.0;
            for (int a = 0; a < d; a++) {
                double delta = -tau * step.get(a, 0);
                vbase[a] += delta;
                stepNorm += delta * delta;
            }
            // Newton converges to the root of the differenced gradient, whose own
            // O(hstep^2) bias puts any absolute gradient target out of reach
            if (FastMath.sqrt(stepNorm) < 1e-10) {
                converged = true;
                break;
            }
        }
        if (!converged) {
            InputOutput.line_warning(InputOutput.mfilename(new Object()),
                    "the saddle point search did not converge, the estimate may be inaccurate.\n");
        }

        Matrix vsad = new Matrix(1, d);
        for (int a = 0; a < d; a++) {
            vsad.set(0, a, vbase[a]);
        }

        // ---- cumulants of the tilted distribution at the saddle ----
        cgf.reset(vbase);
        double K0 = cgf.at(new int[d]);
        Matrix Sigma = new Matrix(d, d);
        for (int a = 0; a < d; a++) {
            for (int b = 0; b < d; b++) {
                Sigma.set(a, b, secondDiff(cgf, d, a, b));
            }
        }
        for (int a = 0; a < d; a++) {
            for (int b = a + 1; b < d; b++) {
                double sym = (Sigma.get(a, b) + Sigma.get(b, a)) / 2;
                Sigma.set(a, b, sym);
                Sigma.set(b, a, sym);
            }
        }
        double minEig = Double.POSITIVE_INFINITY;
        List<Complex> eigs = Sigma.eig();
        for (int i = 0; i < eigs.size(); i++) {
            double re = eigs.get(i).getReal();
            if (re < minEig) {
                minEig = re;
            }
        }
        if (minEig <= 0) {
            throw new RuntimeException("pfqn_nre: the tilted covariance is singular, a class has no demand at any station.");
        }

        double[][][] k3 = new double[d][d][d];
        for (int a = 0; a < d; a++) {
            for (int b = 0; b < d; b++) {
                for (int c = 0; c < d; c++) {
                    double acc = 0;
                    for (int s = 0; s < 8; s++) {
                        int s1 = 1 - 2 * ((s >> 0) & 1);
                        int s2 = 1 - 2 * ((s >> 1) & 1);
                        int s3 = 1 - 2 * ((s >> 2) & 1);
                        int[] off = new int[d];
                        off[a] += s1;
                        off[b] += s2;
                        off[c] += s3;
                        acc += s1 * s2 * s3 * cgf.at(off);
                    }
                    k3[a][b][c] = acc / (8 * HSTEP * HSTEP * HSTEP);
                }
            }
        }

        double[][][][] k4 = new double[d][d][d][d];
        for (int a = 0; a < d; a++) {
            for (int b = 0; b < d; b++) {
                for (int c = 0; c < d; c++) {
                    for (int e = 0; e < d; e++) {
                        double acc = 0;
                        for (int s = 0; s < 16; s++) {
                            int s1 = 1 - 2 * ((s >> 0) & 1);
                            int s2 = 1 - 2 * ((s >> 1) & 1);
                            int s3 = 1 - 2 * ((s >> 2) & 1);
                            int s4 = 1 - 2 * ((s >> 3) & 1);
                            int[] off = new int[d];
                            off[a] += s1;
                            off[b] += s2;
                            off[c] += s3;
                            off[e] += s4;
                            acc += s1 * s2 * s3 * s4 * cgf.at(off);
                        }
                        k4[a][b][c][e] = acc / (16 * HSTEP * HSTEP * HSTEP * HSTEP);
                    }
                }
            }
        }

        // ---- second-order Edgeworth factor, see the class header ----
        Matrix Sinv = Sigma.inv();
        double[][] S = new double[d][d];
        for (int a = 0; a < d; a++) {
            for (int b = 0; b < d; b++) {
                S[a][b] = Sinv.get(a, b);
            }
        }
        double rho4 = 0;
        for (int a = 0; a < d; a++) {
            for (int b = 0; b < d; b++) {
                for (int c = 0; c < d; c++) {
                    for (int e = 0; e < d; e++) {
                        rho4 += k4[a][b][c][e] * S[a][b] * S[c][e];
                    }
                }
            }
        }
        double[] u = new double[d];
        for (int c = 0; c < d; c++) {
            double acc = 0;
            for (int a = 0; a < d; a++) {
                for (int b = 0; b < d; b++) {
                    acc += S[a][b] * k3[a][b][c];
                }
            }
            u[c] = acc;
        }
        double rhoA = 0;
        for (int c = 0; c < d; c++) {
            for (int e = 0; e < d; e++) {
                rhoA += u[c] * S[c][e] * u[e];
            }
        }
        // staged contraction of k3 against three copies of Sigma^{-1}
        double[][][] T1 = new double[d][d][d];
        for (int i = 0; i < d; i++) {
            for (int b = 0; b < d; b++) {
                for (int c = 0; c < d; c++) {
                    double acc = 0;
                    for (int a = 0; a < d; a++) {
                        acc += S[i][a] * k3[a][b][c];
                    }
                    T1[i][b][c] = acc;
                }
            }
        }
        double[][][] T2 = new double[d][d][d];
        for (int i = 0; i < d; i++) {
            for (int j = 0; j < d; j++) {
                for (int c = 0; c < d; c++) {
                    double acc = 0;
                    for (int b = 0; b < d; b++) {
                        acc += S[j][b] * T1[i][b][c];
                    }
                    T2[i][j][c] = acc;
                }
            }
        }
        double rhoB = 0;
        for (int i = 0; i < d; i++) {
            for (int j = 0; j < d; j++) {
                for (int k = 0; k < d; k++) {
                    double acc = 0;
                    for (int c = 0; c < d; c++) {
                        acc += S[k][c] * T2[i][j][c];
                    }
                    rhoB += k3[i][j][k] * acc;
                }
            }
        }
        double corr = 1 + rho4 / 8 - (3 * rhoA + 2 * rhoB) / 24;
        if (corr <= 0) {
            InputOutput.line_warning(InputOutput.mfilename(new Object()),
                    "the Edgeworth correction is non-positive, falling back on the saddlepoint term.\n");
            corr = 1;
        }

        double lGs = K0 - dot(Nd, vbase) - (d / 2.0) * FastMath.log(2 * FastMath.PI)
                - 0.5 * FastMath.log(Sigma.det()) + lGscale;
        return new Result(lGs + FastMath.log(corr), lGs, vsad);
    }

    /** Mixed second difference of the cumulant generating function on the stencil. */
    private static double secondDiff(Cgf cgf, int d, int a, int b) {
        int[] pp = new int[d];
        pp[a] += 1;
        pp[b] += 1;
        int[] pm = new int[d];
        pm[a] += 1;
        pm[b] -= 1;
        int[] mp = new int[d];
        mp[a] -= 1;
        mp[b] += 1;
        int[] mm = new int[d];
        mm[a] -= 1;
        mm[b] -= 1;
        return (cgf.at(pp) - cgf.at(pm) - cgf.at(mp) + cgf.at(mm)) / (4 * HSTEP * HSTEP);
    }

    private static int[] unitoff(int d, int a, int sgn) {
        int[] off = new int[d];
        off[a] = sgn;
        return off;
    }

    private static double dot(double[] x, double[] y) {
        double acc = 0;
        for (int i = 0; i < x.length; i++) {
            acc += x[i] * y[i];
        }
        return acc;
    }

    /**
     * Cumulant generating function of the tilted single-class model, memoised on
     * the finite-difference stencil around the current expansion point.
     */
    private static final class Cgf {
        private final Matrix L;
        private final Matrix Nt;
        private final Matrix alpha;
        private final SolverOptions options;
        private final int d;
        private final Map<Long, Double> cache = new HashMap<Long, Double>();
        private double[] vbase;

        Cgf(Matrix L, Matrix Nt, Matrix alpha, SolverOptions options, int d) {
            this.L = L;
            this.Nt = Nt;
            this.alpha = alpha;
            this.options = options;
            this.d = d;
            this.vbase = new double[d];
        }

        /** Move the expansion point, which invalidates every memoised value. */
        void reset(double[] v) {
            this.vbase = v.clone();
            this.cache.clear();
        }

        /** Value at vbase+off*hstep. */
        double at(int[] off) {
            long key = 0;
            for (int r = d - 1; r >= 0; r--) {
                key = key * SPAN + (off[r] + 4);
            }
            Double cached = cache.get(key);
            if (cached != null) {
                return cached;
            }
            double[] v = new double[d];
            for (int r = 0; r < d; r++) {
                v[r] = vbase[r] + off[r] * HSTEP;
            }
            double y = atPoint(v);
            cache.put(key, y);
            return y;
        }

        /** log of the single-class LLD constant at the class tilt X=[exp(v),1]. */
        double atPoint(double[] v) {
            int M = L.getNumRows();
            int R = L.getNumCols();
            Matrix Lx = new Matrix(M, 1);
            for (int i = 0; i < M; i++) {
                double acc = 0;
                for (int r = 0; r < R; r++) {
                    double xr = r < d ? FastMath.exp(v[r]) : 1.0;
                    acc += L.get(i, r) * xr;
                }
                Lx.set(i, 0, acc);
            }
            return Pfqn_lldsingle.pfqn_lldsingle(Lx, Nt, alpha, options).lG;
        }
    }
}
